### ISSUE TITLE
[AAS Part 2] Custom Information in ServiceDescription

### ISSUE THREAD

**Issue by mjacoby** [admin-shell-io/aas-specs-api#163]
In AAS Part 2 (IDTA Number `01002-3-0`) section `10.2.6.`  p.85 Definition of `ServiceDescription` it says
```
Additional, proprietary attributes might be included
```

This seems like a bad idea as it would result in a unknown/non-deterministic result structure of the `GetSelfDescription` operation. This not only makes implementation harder as deserialization of such an object can only return a generic, non-AAS-specific class but alsois exactly the reason the content modifiers are now part of the URL path instead of the query string.

This is an example of a valid payload illustrating the problem
```JSON
{
	"profiles": [
		...
	],
	"foo": {
		"bar": ...
	}	
}
```

As I think having custom data here makes sense, my suggestion is to provide a deterministic structure to hold that kind of data. There seem to be (at least) two type of data that should be treated differently

- custom capabilities, e.g. like supporting an additional, proprietary content modifier (in form of string[])
- custom attributes, all other custom attributes that are not capabilites, e.g. server timestamp oder version (in form of a key-value map)

```JSON
{
	"profiles": [
		...
	],
	"custom-capabilities": [
		"custom-capability-A",
		"custom-capability-B",
		...
	],
	"custom-attributes": {
		"server-timestamp": "2023-05-15 12:00:00Z",
                ...
	}
}
```

**Comment by sebbader-sap** [comment]
@mjacoby please present this topic in the next group meeting if it still exists.

**Comment by mjacoby** [comment]
Yes, issue still exists. Please put it on the agenda for next call, I will present it.

**Comment by sebbader-sap** [comment]
The topic grew a bit. As we recognized, the JSON schemas (metamodel + API) are _open_, while XML is _closed_ as defined through its schema. We therefore have a missmatch:
1. XML: Non-standardised properties must go via `extenstions` only, any other approach leads to validation errors.
2. API + JSON: Non-standardised properties can be added everywhere, as the schema/OpenAPI doesn't care at all

This means in particular that conversions between the data formats lead to information loss, if the behavior is defined at all.

**Comment by BirgitBoss** [comment]
I think the formulation "Additional, proprietary attributes might be included" is misleading. The idea is that the enumeration with the profiles is an open enumeration and additional proprietary profile can be added:
https://industrialdigitaltwin.io/aas-specifications/IDTA-01002/v3.1/specification/interfaces-payload.html#ServiceSpecificationProfileEnum

So I propose to reformulate the explanation of ServiceDescription from (substitute attribute with profile):
`The self-describing information of an API Implementation. It enables servers to present their capabilities to the clients, in particular which profiles they implement. At least one defined profile is required. Additional, proprietary attributes might be included. Nevertheless, the server must not expect that a regular client understands them.`
to
`The self-describing information of an API Implementation. It enables servers to present their capabilities to the clients, in particular which profiles they implement. At least one defined profile is required. Additional, proprietary profiles might be included. Nevertheless, the server must not expect that a regular client understands them.`

Additionally we might think of a generic way to specify that an enumeration is open or closed.

**Comment by sebbader-sap** [comment]
Please check the AAS Workstream decision in https://github.com/admin-shell-io/aas-specs-metamodel/issues/622

**Comment by mristin** [comment]
Please apologize for jumping in so late; the issue was not on my radar. Please make the schema closed (`additionalProperties: false` everywhere). Open schemas will create a huge mess in the ecosystem as people start putting their own custom-tailored stuff. This will definitely kill interoperability.

Moreover, objects passed from different versions of the meta-model will simply go unnoticed and the data will be lost in the process without anybody noticing -- the attributes from a different version will simply be "swallowed" without any further warning.

**Comment by alexgordtop** [comment]
@mristin Following your statement we wouldn't be able to add any new attribute to the JSON schema without causing serious trouble with older versions.

For example:
We plan to introduce optional attributes like "creationDate" oder "updateDate" to the AdministrativeInformation with version 3.2 (don't nail me down on the exact naming). That is considered a *non-breaking* change, since new clients will use the attributes, while old clients get not irritated because of the "additionalProperties: true" principal.

From my perspective a major advantage from json over xml and probably one of the reasons, why nowadays xml isn't used as much as compared to 2010 and the whole WS* ecosystem.

And I'm not alone with that perspective:
> While Robert C. Martin (Uncle Bob) does not explicitly write about HTTP API backward compatibility, his Open/Closed Principle (OCP) provides a strong theoretical foundation for it.
> 
> The Open/Closed Principle states that:
> 
> Software entities should be open for extension, but closed for modification.
> 
> Applied to APIs, this means:
> 	•	Adding new optional attributes is an extension of the contract and therefore non-breaking
> 	•	Removing or changing existing required attributes modifies the contract and is therefore breaking
> 
> In other words, APIs should be designed so that new capabilities can be added without forcing existing clients to change. This aligns directly with the common API guideline that additive, optional changes are backward compatible, while destructive changes are not.
> 
> Although OCP originated with Bertrand Meyer, Uncle Bob popularized it as part of SOLID, and it is widely used to justify backward-compatible API evolution.
> 
> Conclusion:
> Uncle Bob’s Open/Closed Principle supports the industry convention that adding new optional fields is non-breaking, while removing or changing mandatory fields constitutes a breaking change.

=> https://blog.cleancoder.com/uncle-bob/2014/05/12/TheOpenClosedPrinciple.html

**Comment by mristin** [comment]
@alexgordtop  thanks for giving the background! 

Bertrand Meyer introduced the Open-Closed Principle based on Barbara Liskov's Behavioral Subtyping. This presupposes *abstract* data types (ADTs), implying polymorphism and dynamic dispatch. 

The current proposal considers, hower, *concrete* data types. What it actually amounts to is *object slicing*, where members of a descendant structure are ignored during a copy, breaking the polymorphism. Here is a code snippet illustrating that behavior in C++: 

```cpp
struct A {
    int x;
}; 

struct B : A {
    int y;
}; 

B b;
b.x = 1;
b.y = 2; 

A a = b;  // object slicing
``` 

`A a = b;` creates a new `A` object. Only the `A` portion of b is copied. The polymorphism is not preserved when copying by value. 

Alternatively, you can also think of the server as a container. The current proposal wants to introduce co-variance on writing, which again breaks the conventional type systems. In all major languages -- unless both reads and writes are invariant -- reads are co-variant, but writes are contra-variant. 

Consider another code snippet in pseudo-code:
```
type Animal
type Cat extends Animal
type Dog extends Animal 

type Box<T> {
    write(T value)
    read() -> T
} 

Box<Cat> catBox = new Box<Cat>() 

// If writes were covariant (the mistake):
Box<Animal> animalBox = catBox 

// ... then this looks legal to the type system:
animalBox.write(new Dog()) 

// ... but now:
Cat c = catBox.read()   
// 💥 Dog where a Cat is expected
``` 

Hence, as soon as you have writes, you have to stick to the compatible less-or-equal meta-model version, not greater-equal.

In other words, once you allow writes, the client must restrict itself to a meta-model version no newer than what the server guarantees it accepts. Otherwise you silently drop data (object slicing), which breaks type soundness, and ensues in catastrophic failures. 

*edit: typos*
